File: data_sync.API/Services/SyncStateService.cs
namespace data_sync.API.Services;

/// <summary>
/// Verwaltet den letzten bekannten Synchronisationsstand einer Datei.
/// </summary>
public class SyncStateService
{
    private readonly MariaDbService _mariaDbService;

    public SyncStateService(MariaDbService mariaDbService)
    {
        _mariaDbService = mariaDbService;
    }

    /// <summary>
    /// Legt den letzten Synchronisationsstand für eine Datei an oder aktualisiert ihn.
    /// </summary>
    public async Task UpdateSyncStateAsync(string filePath, string hashValue, long fileSize)
    {
        await using var connection = await _mariaDbService.OpenConnectionAsync();
        if (connection == null)
        {
            throw new InvalidOperationException("MariaDB connection could not be established.");
        }

        await using var dbCommand = connection.CreateCommand();
        dbCommand.CommandText = @"
            INSERT INTO LastSyncState (file_path, hash_value, file_size)
            VALUES (@filePath, @hashValue, @fileSize)
            ON DUPLICATE KEY UPDATE
                hash_value = @hashValue,
                file_size = @fileSize,
                last_synced_at = CURRENT_TIMESTAMP;";

        dbCommand.Parameters.AddWithValue("@filePath", filePath);
        dbCommand.Parameters.AddWithValue("@hashValue", hashValue);
        dbCommand.Parameters.AddWithValue("@fileSize", fileSize);

        await dbCommand.ExecuteNonQueryAsync();
    }

    /// <summary>
    /// Entfernt den letzten Synchronisationsstand für eine Datei.
    /// </summary>
    public async Task DeleteSyncStateAsync(string filePath)
    {
        await using var connection = await _mariaDbService.OpenConnectionAsync();
        if (connection == null)
        {
            throw new InvalidOperationException("MariaDB connection could not be established.");
        }

        await using var dbCommand = connection.CreateCommand();
        dbCommand.CommandText = "DELETE FROM LastSyncState WHERE file_path = @filePath;";
        dbCommand.Parameters.AddWithValue("@filePath", filePath);

        await dbCommand.ExecuteNonQueryAsync();
    }
}

File: data_sync.FLUTTER/lib/services/sync_state_service.dart
import '../handlers/sqlite_handler.dart';
import '../models/sync_models.dart';

/// Verwaltet den letzten erfolgreichen Sync-Zustand lokaler Dateien.
class SyncStateService {
  SyncStateService({SqliteHandler? sqliteHandler})
    : _sqliteHandler = sqliteHandler ?? SqliteHandler();

  final SqliteHandler _sqliteHandler;

  /// Lädt alle bekannten Sync-Zustände aus SQLite.
  Future<Map<String, SyncStateEntry>> getAllStates() async {
    final results = await _sqliteHandler.executeQuery(
      '''
      SELECT file_name, file_path, file_size, hash_value, created_at, last_modified
      FROM last_sync_state
      ''',
    );

    final states = <String, SyncStateEntry>{};

    for (final row in results) {
      final entry = SyncStateEntry.fromMap(row);
      states[entry.relativePath] = entry;
    }

    return states;
  }

  /// Legt einen Sync-Zustand an oder aktualisiert ihn.
  Future<void> upsertState(SyncStateEntry entry) async {
    await _sqliteHandler.executeQuery(
      '''
      INSERT INTO last_sync_state (
        file_name,
        file_path,
        file_size,
        hash_value,
        created_at,
        last_modified
      ) VALUES (?, ?, ?, ?, ?, ?)
      ON CONFLICT(file_path) DO UPDATE SET
        file_name = excluded.file_name,
        file_size = excluded.file_size,
        hash_value = excluded.hash_value,
        created_at = excluded.created_at,
        last_modified = excluded.last_modified
      ''',
      params: [
        entry.fileName,
        entry.relativePath,
        entry.size,
        entry.sha256,
        entry.createdAt.toUtc().toIso8601String(),
        entry.lastModified.toUtc().toIso8601String(),
      ],
      returnResult: false,
    );
  }

  /// Entfernt einen Sync-Zustand nach einer bestätigten Löschung.
  Future<void> deleteState(String relativePath) async {
    await _sqliteHandler.executeQuery(
      'DELETE FROM last_sync_state WHERE file_path = ?',
      params: [relativePath],
      returnResult: false,
    );
  }
}
